sklearn/cluster/_optics.py

```diff
@@ -603,15 +612,15 @@ def _set_reach_dist(
     # Only compute distances to unprocessed neighbors:
     if metric == "precomputed":
         dists = X[point_index, unproc]
+        if issparse(dists):
+            dists = dists.toarray().ravel()
```

Comment:
I do not think we need to make this dense:

```suggestion
            dists.sort_indices()
            dists = dists.data
```

It is a trade off between sorting the indices and densifying the sparse matrix. In the sparse precomputed metric use case, I suspect there are not many indices and densifying is more expensive.